src/MadeTech/CacheWarming/CacheWarmer.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+namespace MadeTech\CacheWarming;
+
+
+use MadeTech\CacheWarming\Config\HrefLangExpansion;
+
+class CacheWarmer
+{
+    /** @var UseCase\WarmUpCacheForSiteMap */
+    private $warmUpCacheForSite;
+
+    /** @var Config\ConfigProvider */
+    private $configuration;
+
+    /** @var UrlRetriever */
+    private $urlRetriever;
+
+    /** @var string[] */
+    private $htmlPages = [];
+
+    public function __construct(
+        UseCase\WarmUpCacheForSiteMap $warmUpCacheForSite,
+        Config\ConfigProvider $configuration,
+        UrlRetriever $urlRetriever
+    ) {
+        $this->warmUpCacheForSite = $warmUpCacheForSite;
+        $this->configuration = $configuration;
+        $this->urlRetriever = $urlRetriever;
+    }
+
+    public function warmCaches(UseCase\WarmUpCacheForSitePresenter $presenter)
+    {
+        $siteMaps = $this->getSiteMaps();
+        foreach ($siteMaps as $siteMap) {
+            $this->warmUpCacheForSite->warmUpSiteCache($siteMap, $presenter);
+        }
+    }
+
+    /**
+     * @param HrefLangExpansion $hrefLangExpansion
+     *
+     * @return string[]
+     */
+    private function getAlternateUrls(HrefLangExpansion $hrefLangExpansion)
+    {
+        $alternateUrls = [];
+        foreach ($this->getLinkTagsFromSite($hrefLangExpansion) as $linkTag) {
+            $attributes = $this->getLinkTagAttributes($linkTag);
+            if ($this->isAlternate($attributes)) {
+                $alternateUrls[] = (string)$attributes['href']->value;
+            }
+        }
+
+        return $alternateUrls;
+    }
+
+    private function getHrefLangExpansionSiteMaps(HrefLangExpansion $hrefLangExpansion)
+    {
+        return array_map(
+            function ($url) use ($hrefLangExpansion) {
+                return preg_replace(
+                    $hrefLangExpansion->getRegexMatchingString(),
+                    $hrefLangExpansion->getRegexReplacingString(),
+                    $url
+                );
+            },
+            $this->getAlternateUrls($hrefLangExpansion)
+        );
+    }
+
+    private function getSiteHtml(HrefLangExpansion $hrefLangExpansion)
+    {
+        $url = $hrefLangExpansion->getSiteUrl();
+        if (!isset($this->htmlPages[$url])) {
+            $this->htmlPages[$url] = $this->urlRetriever->get($url);
+        }
+
+        return $this->htmlPages[$url];
+    }
+
+    private function getDomDocument(HrefLangExpansion $hrefLangExpansion)
+    {
+        $dom = new \DomDocument();
+        @$dom->loadHTML($this->getSiteHtml($hrefLangExpansion), LIBXML_NOWARNING);
```

Comment:
is libxml good enough to parse broken html or should we use something dedicated?